pkg/cluster/acsk/action/action_utils.go

```diff
@@ -0,0 +1,345 @@
+// Copyright © 2018 Banzai Cloud
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package action
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"strings"
+	"time"
+
+	aliErrors "github.com/aliyun/alibaba-cloud-sdk-go/sdk/errors"
+	"github.com/aliyun/alibaba-cloud-sdk-go/sdk/requests"
+	"github.com/aliyun/alibaba-cloud-sdk-go/services/cs"
+	"github.com/aliyun/alibaba-cloud-sdk-go/services/ess"
+	"github.com/banzaicloud/pipeline/model"
+	"github.com/banzaicloud/pipeline/pkg/cluster/acsk"
+	pkgErrors "github.com/banzaicloud/pipeline/pkg/errors"
+	"github.com/goph/emperror"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+)
+
+func deleteCluster(log logrus.FieldLogger, clusterID string, csClient *cs.Client) error {
+
+	if len(clusterID) == 0 {
+		return nil
+	}
+
+	_, err := waitUntilClusterCreateOrScaleComplete(log, clusterID, csClient, true)
+	if err != nil {
+		log.Warn("Error happened during waiting for cluster state to be deleted ", err)
+		return emperror.Wrap(err, "could not delete cluster!")
+	}
+
+	req := cs.CreateDeleteClusterRequest()
+	req.ClusterId = clusterID
+	req.SetScheme(requests.HTTPS)
+	req.SetDomain(acsk.AlibabaApiDomain)
+
+	resp, err := csClient.DeleteCluster(req)
+	if err != nil {
+		if sdkErr, ok := err.(*aliErrors.ServerError); ok {
+			if strings.Contains(sdkErr.Message(), "ErrorClusterNotFound") {
+				// Cluster has been already deleted
+				return nil
+			}
+		}
+		return errors.WithMessage(err, fmt.Sprintf("DeleteClusterResponse: %#v \n", resp.BaseResponse))
+	}
+
+	if resp.GetHttpStatus() != http.StatusAccepted {
+		return fmt.Errorf("unexpected http status code: %d", resp.GetHttpStatus())
+	}
+
+	return nil
+}
+
+func describeScalingInstances(essClient *ess.Client, asgId, scalingConfId, regionId string) (*ess.DescribeScalingInstancesResponse, error) {
+	describeScalingInstancesRequest := ess.CreateDescribeScalingInstancesRequest()
+	describeScalingInstancesRequest.SetScheme(requests.HTTPS)
+	describeScalingInstancesRequest.SetDomain("ess." + regionId + ".aliyuncs.com")
```

Comment:
The api endpoint could be extracted into a const similar to `ossEndpointFmt = "https://oss-%s.aliyuncs.com"`